v2/netconf/server/ssh/trace.go

```diff
@@ -0,0 +1,123 @@
+package ssh
+
+import (
+	"context"
+	"log"
+	"net"
+
+	"github.com/imdario/mergo"
+)
+
+// unique type to prevent assignment.
+type sshEventContextKey struct{}
+
+// ContextSshTrace returns the Trace associated with the
+// provided context. If none, it returns nil.
+func ContextSshTrace(ctx context.Context) *Trace {
+	trace, _ := ctx.Value(sshEventContextKey{}).(*Trace)
+	if trace == nil {
+		trace = NoOpLoggingHooks
+	} else {
+		_ = mergo.Merge(trace, NoOpLoggingHooks) // nolint: gosec, errcheck
+	}
+	return trace
+}
+
+// WithSshTrace returns a new context based on the provided parent
+// ctx. Requests made with the returned context will use
+// the provided trace hooks
+func WithSshTrace(ctx context.Context, trace *Trace) context.Context {
+	ctx = context.WithValue(ctx, sshEventContextKey{}, trace)
+	return ctx
+}
+
+// Trace defines a structure for handling trace events
+type Trace struct {
```

Comment:
Does the type need callbacks for teardown?

Reply:
At some stage yes - see my comment on the pull request : _diagnostic reporting/trace hooks need to be extended/refined_.   
But I'd rather do that when I've been using the framework and have a better idea as to what is useful.